// Copyright Titanium I.T. LLC. License granted under terms of "The MIT License."
"use strict";

const { suite, assert } = require("tests");
const TestRunner = require("./test_runner");
const path = require("node:path");
const TestSuite = require("./test_suite");
const TestResult = require("tests/test_result");
const fs = require("node:fs/promises");
const Clock = require("infrastructure/clock");

module.exports = suite(({ beforeEach, it }) => {

	let TEST_MODULE_PATH;
	beforeEach(async ({ getConfig }) => {
		const testDir = getConfig("scratchDir");

		TEST_MODULE_PATH = `${testDir}/_test_runner_module.js`;
		await deleteTempFilesAsync(testDir);
	});

	it("runs test modules", async () => {
		const { runner } = create();
		await writeTestModuleAsync(`// passes`);

		const results = await runner.runIsolatedAsync([ TEST_MODULE_PATH ]);

		const expectedResult = TestResult.suite([], [
			TestResult.suite([], [
				TestResult.pass("test", TEST_MODULE_PATH)
			], TEST_MODULE_PATH),
		]);

		assert.objEqual(results, expectedResult);
	});

	it("passes through config", async () => {
		const myConfig = { myConfig: "my_config" };
		const { runner } = create();

		await writeTestModuleAsync(`assert.fail(getConfig("myConfig"));`);
		const results = await runner.runIsolatedAsync([ TEST_MODULE_PATH ], { config: myConfig });

		assertFailureMessage(results, "my_config");
	});

	it("notifies caller of completed tests", async () => {
		const { runner } = create();

		const progress = [];
		const notifyFn = result => progress.push(result);

		await writeTestModuleAsync(`// passes`);
		await runner.runIsolatedAsync([ TEST_MODULE_PATH ], { notifyFn });

		assert.deepEqual(progress, [
			TestResult.pass("test", TEST_MODULE_PATH),
		]);
	});

	it("does not cache test modules from run to run", async () => {
		const { runner } = create();

		await writeTestModuleAsync(`assert.fail("module was cached, and shouldn't have been");`);
		await runner.runIsolatedAsync([ TEST_MODULE_PATH ]);

		await writeTestModuleAsync(`assert.fail("module was not cached");`);
		const results = await runner.runIsolatedAsync([ TEST_MODULE_PATH ]);

		assertFailureMessage(results, "module was not cached");
	});

	it("isolates tests", async () => {
		const { runner } = create();

		await writeTestModuleAsync(`global._test_runner_test = true;`);
		await runner.runIsolatedAsync([ TEST_MODULE_PATH ]);

		await writeTestModuleAsync(`assert.fail("global should be undefined: " + global._test_runner_test);`);
		const results = await runner.runIsolatedAsync([ TEST_MODULE_PATH ]);

		assertFailureMessage(results, "global should be undefined: undefined");
	});

	it("supports process.chdir(), which isn't allowed in Worker threads", async () => {
		const { runner } = create();

		await writeTestModuleAsync(`
			process.chdir(".");
			assert.fail("process.chdir() should execute without error");
		`);
		const results = await runner.runIsolatedAsync([ TEST_MODULE_PATH ]);

		assertFailureMessage(results, "process.chdir() should execute without error");
	});

	it("handles uncaught promise rejections", async () => {
		const { runner } = create();

		await writeTestModuleAsync(`Promise.reject(new Error("my error"));`);
		const results = await runner.runIsolatedAsync([ TEST_MODULE_PATH ]);

		assert.deepEqual(results, TestResult.suite([], [
			TestResult.fail("Unhandled error in tests", new Error("my error")),
		]));
	});

	it("handles infinite loops", async () => {
		const { runner, clock } = create();

		await writeTestModuleAsync(`while (true);`);
		const resultsPromise = runner.runIsolatedAsync([ TEST_MODULE_PATH ]);

		await clock.tickAsync(TestSuite.DEFAULT_TIMEOUT_IN_MS);

		assert.deepEqual(await resultsPromise, TestResult.suite([], [
			TestResult.fail("Test runner watchdog", "Detected infinite loop in tests"),
		]));


		// It also doesn't fail when tests run long (without an infinite loop), but I don't know how to test that.
		// Probably need a nullable process?
	});


	function assertFailureMessage(results, expectedFailure) {
		assert.equal(results.suite[0].suite[0].error.message, expectedFailure);
	}

	async function writeTestModuleAsync(bodySourceCode) {
		await fs.writeFile(TEST_MODULE_PATH, `
			"use strict";
			
			const { suite, assert } = require` + `("${path.resolve(__dirname, "../tests")}");
			
			module.exports = suite(({ it }) => {
				it("test", ({ getConfig }) => {
					${bodySourceCode}
				});
			});
		`);
	}

	async function deleteTempFilesAsync(testDir) {
		assert.defined(testDir);
		await fs.rm(testDir, { recursive: true, force: true });
		await fs.mkdir(testDir, { recursive: true });
	}

});

function create({
	clock = Clock.createNull(),
} = {}) {
	const runner = new TestRunner(clock);

	return { runner, clock };
}